nmt.go

```diff
@@ -0,0 +1,142 @@
+package nmt
+
+import (
+	"bytes"
+	"crypto"
+	"errors"
+	"fmt"
+
+	"github.com/liamsi/merkletree"
```

Comment:
This repo should be moved to lazyledger org

Reply:
It's a fork of the NebolousLabs' tree with very few changes. Ideally, we can merge these changes into their library. If that doesn't happen, I'll move it to the LL org 👍  
(see: https://gitlab.com/NebulousLabs/merkletree/-/issues/21)

Reply:
There are other changes that we are likely going to need to make to the underlying merkletree implementation that may not be mergeable upstream. For example, we should probably change the implementation to store key=>value pairs of nodes in the tree using the same interface as the SMT MapStore, so that we can plug in other MapStores like an IPFS-based one, and reuse the same code: https://github.com/lazyledger/smt-ipfs-mapstore

This is something that needs to be investigated a bit though, as from my understanding the NebulousLabs merkletree library I think may compute the root on demand and doesn't keep around any of the old nodes in memory. Alternatively, it may be possible to implement this MapStore storage functionality within the NMT library, which uses the MapStore as "cache", and only invokes the NebulousLabs merkletree library to recompute roots if the node data is unavailable in the MapStore.

Reply:
> This is something that needs to be investigated a bit though, as from my understanding the NebulousLabs merkletree library I think may compute the root on demand and doesn't keep around any of the old nodes in memory.

That's my understanding too. An alternative could be to look further into trillian, it allows to recompute the tree quickly [in memory](https://github.com/google/trillian/blob/7166f485ebc965d24116db667c9e058a8ed9275c/merkle/memory_merkle_tree.go#L69) (I did some measurements and it seemed a bit faster then NebolousLabs') but also comes with a storage backend where you can query for ranges etc: https://github.com/google/trillian/blob/master/storage/log_storage.go

My plan for now was to abstract storage away and give the prover/verifier some kind of leaf iterator.